R/spread_with_duplicates.R

```diff
@@ -0,0 +1,177 @@
+#'   columns. If only one value exists per key, then this functions is the same
+#'   as \code{\link[tidyr]{spread}}. Otherwise, aggregation function
+#'   \code{aggfunc} is applied. If no function is passed, then the key-value(s)
+#'   are spread on multiple rows.
+
+#' @param data A dataframe.
+#' @param key,value Column names or positions. This is passed to
+#'   \code{\link[tidyselect]{vars_pull}}. These arguments are passed by
+#'   expression and support \code{\link[rlang]{quasiquotation}} (you can unquote
+#'   column names or column positions)
+#' @param fill If set, missing values will be replaced with this value.
+#' @param convert If \code{TRUE}, \code{\link[utils]{type.convert}} with
+#'   \code{asis = TRUE} will be run on each of the new columns. This is useful
+#'   if the value column was a mix of variables that was coerced to a string. If
+#'   the class of the value column was factor or date, note that will not be
+#'   true of the new columns that are produced, which are coerced to character
+#'   before type conversion.
+#' @param drop If \code{FALSE}, will keep factor levels that don't appear in the
+#'   data, filling in missing combinations with \code{fill}.
+#' @param sep If \code{NULL}, the column names will be taken from the values of
+#'   \code{key} variable. If non-\code{NULL}, the column names will be given by
+#'   "<key_name><sep><key_value>".
+#' @param aggfunc Aggregation function. Default: NA (keep dulicates)
+#'
+#' @return A data.frame.
+#'
+#' @export
+#'
+#' @importFrom purrr map map2 reduce compact
+#' @importFrom magrittr %<>%
+#' @importFrom rlang sym syms is_character
+#' @importFrom dplyr summarize funs rename mutate_all mutate_at filter full_join
+#'   pull %>% one_of group_by vars
+#' @importFrom tidyselect vars_pull enquo
+#' @importFrom tidyr complete_
+#'
+#' @examples
+#' \dontrun{
+#' test1 <- data.frame(
+#' col1 = c(1, 1, 1, 1),
+#' col2 = c("H", "H", "H", "H"),
+#' key = c("A", "B", "C", "C"),
+#' value = c("R", "S", "T", "X"),
+#' stringsAsFactors = FALSE
+#' )
+#' spread_with_duplicates(test1, key, value)
+#' spread_with_duplicates(test1, 3, 4)
+#' spread_with_duplicates(test1, -2, -1)
+#' spread_with_duplicates(test1, "key", "value")
+#'
+#' # with NAs
+#' test2 <- data.frame(
+#' col1 = c(1, 1, 1, 2),
+#' key = c("A", "C", "C", "A"),
+#' value = c("R", "T", "X", "R"),
+#' stringsAsFactors = FALSE
+#' )
+#' spread_with_duplicates(test2, key, value)
+#' spread_with_duplicates(test2, key, value, fill = "No idea")
+#'
+#' # apply aggregate function
+#' test3 <- data.frame(
+#' col1 = c(1,1,1,1),
+#'   col2  = c("H", "H", "H", "H"),
+#'   key = c("A", "B", "C", "C"),
+#'   value = c(2, 3, 1, 8),
+#'   stringsAsFactors = FALSE
+#' )
+#' spread_with_duplicates(test1, key, value, aggfunc = str_c, collapse = "-")
+#' spread_with_duplicates(test2, key, value, aggfunc = min)
+#' spread_with_duplicates(test2, key, value, aggfunc = mean)
```

Comment:
I would provide another example with numeric values instead of characters, as currently this returns all NA values and provides the warning:

```
> spread_with_duplicates(test2, key, value, aggfunc = mean)
# A tibble: 2 x 3
   col1     A     C
  <dbl> <dbl> <dbl>
1    1.    NA    NA
2    2.    NA    NA
Warning messages:
1: In mean.default("R") : argument is not numeric or logical: returning NA
2: In mean.default("R") : argument is not numeric or logical: returning NA
3: In mean.default(c("T", "X")) :
  argument is not numeric or logical: returning NA
```

Reply:
I now see the issue: you reused test2 instead of 3; fixed by 4ed15ef5be1c4ca1a684759d12236ec53729aca9

Reply:
Thanks. I was removing variables which were not anymore used so I tried to rename the `test_x` variables and something went wrong.